gpii/node_modules/solutionsRegistry/src/js/SolutionsUtils.js

```diff
@@ -0,0 +1,107 @@
+/*
+ * GPII Solutions Utils - Independent helper functions and utilities for working with
+ * solutions and their schemas.
+ *
+ * Copyright 2019 Raising the Floor - US
+ *
+ * Licensed under the New BSD license. You may not use this file except in
+ * compliance with this License.
+ *
+ * The research leading to these results has received funding from the European Union's
+ * Seventh Framework Programme (FP7/2007-2013)
+ * under grant agreement no. 289016.
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/GPII/universal/blob/master/LICENSE.txt
+ */
+"use strict";
+var fluid = require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+
+fluid.registerNamespace("gpii.universal.solutionsRegistry");
+
+/**
+ * This function will take a schema from a supported setting in the solutions
+ * registry and return metadata on it's default value, including the type of
+ * the default value. The function will always return a block of json in order
+ * to avoid any ambiguities for a situation where the default value is perhaps
+ * `null`. A key `hasDefault` will be `true` or `false` to communicate this.
+ * If the schema does have a default value, then the properties `type` and
+ * `default` can also be included. Note that there may not always be a `type`
+ * that can be included, the inclusion of this in the return if optional and
+ * will be included if a `type` is declared.
+ *
+ * This is NOT meant to be an exhaustive coverage of all the possible ways
+ * default values can be set for any JSON Schema. We merely want to cover the
+ * cases in use in the solutions registry, focussing on single primitive values,
+ * and situations where we next one level deep due to a requirement of having a
+ * `properties` entry in the schema.
+ *
+ * @param {Object} schema - Schema to search for a default value. This is what
+ * would be in the solutions registry under `supportedSettings` and a
+ * configuration. An example from the Windows Mouse Keys MaxSpeed setting is:
+ * ```json
+ * {
+ *     "title": "Mouse keys speed",
+ *     "description": "Speed of mouse keys",
+ *     "type": "number",
+ *     "multipleOf": 10,
+ *     "default": 80
+ * }
+ * ```
+ * @return {Object} Always returns a json block. For the example above the
+ * return value would be:
+ * ```json
+ * {
+ *     "hasDefault": true,
+ *     "type": "number",
+ *     "default": 80
+ * }
+ * ```
+ * If a default value cannot be detected the return payload will be:
+ * ```json
+ * {
+ *     "hasDefault": false
+ * }
+ * ```
+ */
+gpii.universal.solutionsRegistry.findDefaultValue = function (schema) {
```

Comment:
It's helpful, but it is not used by anything in the codebase - does it belong in the capture tool instead?
Also note that the one place where we do fetch setting defaults, we do not use this approach - https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/PSPChannel.js#L139 - instead, we read the value from build/schemas/solution-schema-codex.json . What do you think, @cindyli , @the-t-in-rtf  ?

Reply:
I was confused as to where and why you'd use this, but then I looked at some of the deeper structures like the SPI settings here:

https://github.com/GPII/universal/blob/76ad0e6d5eb7023a566bcabbc32695e5e844ba00/testData/solutions/win32.json5#L6324

I guess my questions are, what do you expect to do with this potential "deep default" and why do you preserve the "type" but not anything else?  If you're putting up controls or anything else, you don't have enough information to create controls.  You don't know the range any longer, you don't know if the field is required, et cetera.  If you're getting that info somewhere else, why not get the type from there as well?

Also, your code is focused on the `value` property, there are others, like [this object that has RGB properties](
https://github.com/GPII/universal/blob/76ad0e6d5eb7023a566bcabbc32695e5e844ba00/testData/solutions/win32.json5#L5082).  How do you plan to handle those?

It seems to me like you need proper handling for nested properties, and this type of recursion should be possible in both javascript and handlebars templates.  But maybe I just don't get what you're doing, and an example of what you're feeding into this and where you use it would help.  Apologies if you've already covered that in the longer conversation.

I also am not sure this belongs in this particular part of the Solutions Registry.  The use case would help confirm, but if it's truly about UI generation, it seems like this would be an early utility that would make up part of the "presentation registry" instead.

Reply:
@amb26 Well, it's first use is the capture tool (and I've already temporarily duplicated it there), but it will all need to be used in the PPT, Morphic Vaults and other user facing applications build on top of universal.  Quite a number of our UX specs at some point highlight what the default value of a particular setting is.

Also, regardless of the input, whether it's from win32.json or from solution-schema-codex.json, I believe we have the same problems. (and this impl I believe should still work if you pass in the corresponding property/schema either way.  Here is a section from my locally generated `solution-schema-codex.json`:

```json
       "SlowKeysInterval": {
        "title": "Slow keys interval",
        "description": "Slow keys interval time in milliseconds",
        "properties": {
          "path": {
            "type": "string",
            "required": true
          },
          "value": {
            "required": true,
            "type": "integer",
            "minimum": 0,
            "maximum": 20000,
            "default": 1000
          }
        }
      },
      "BounceKeysInterval": {
        "title": "Bounce keys interval",
        "description": "Bounce keys interval time in milliseconds",
        "type": "integer",
        "default": 0
      }
```

Here, you can see that the `default` is at the top level for `BounceKeysInterval`, and then for `SlowKeysInterval` you need to drill down a level to get the default value.

Reply:
@the-t-in-rtf An example of what I'm doing with the `default` value (regardless of whether it's in the top level of the items schema, or further down), is here in the capture tool where you can filter values on and off, depending if they match the default:

<img width="618" alt="image" src="https://user-images.githubusercontent.com/361700/73694314-a5d84000-468c-11ea-8249-417822225f56.png">

In the PPT, on the tables that show settings for each solution there is a column for the default.

Current I'm just preserving `type` because it's the amount of information I need to build these UI's (maybe not a great answer, but the truth). I just need the default, and it's type so I can compare them.  I've haven't yet dealt with RGB values or anything, as the comments on the method allude to, I'm trying to knock out the largest number of settings at the moment that show up on these UI's with the simplest solution so far.  

I'm ok with it being anywhere in universal, but it would be nice to have it in at least one of our most central repos... at the moment every one of our user facing projects still includes universal.  Also, I can imagine a time soon where this default value information would be needed to create some time of data feed endpoint in universal.

Reply:
I still don't understand why you can't just use the schema in its entirety.   We should discuss in the meeting today.